### DISCUSSION TITLE
Support `Date` as JSON-serializable runtime type

### DISCUSSION THREAD

**Discussion by ctjlewis** [vercel/next.js#45744] [Ideas]
### Describe the feature you'd like to request

Next does not recognize Dates as JSON-serializable, yet there is no issue doing this:

```ts
JSON.stringify(new Date())
// "2023-02-08T22:53:19.222Z"
```

To workaround this in `getStaticProps()` etc, you can simply prevent Next from doing this check:

```ts
export async function getServerSideProps(context) {
  // ...
  return {
    props: {
       ...props,
       date: JSON.parse(JSON.stringify(props.date)),
    },
  };
}
```

### Describe the solution you'd like

Simply mark Dates as JSON-serializable.

Both `Date.parse()`/`JSON.parse(new Date())` for ISO timestamps (on client) and `Date.toJSON()` on Node are well-supported since ~2011:  

- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/parse
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toJSON

### Describe alternatives you've considered

The `JSON.parse(JSON.stringify(...))` method above works.

**Comment by joulev** [comment]
`JSON.parse(JSON.stringify(new Date())` returns a string. So it is **not** the same as `new Date()`. The following does **not** work.

```tsx
export default function Page({ date }) {
  return <div>{date.getFullYear()}</div>;
}

export function getStaticProps() {
  return {
    date: JSON.parse(JSON.stringify(new Date())),
  };
}
```

So if your `parse(stringify())` method was to be implemented, it would create inconsistency where you expect a date to be of type `Date` but in fact it is of type `string`, and these two types are obviously not interchangeable. `Date` type magically becoming `string` without explicitly casting by the developer (your suggestion) is *far* worse than `Date` type not being supported (current status), if you ask me.

Instead, a better proposal would be to use a transformer, for example `superjson`. It doesn't seem like the team is interested in that though. [You can follow this guide instead](https://github.com/blitz-js/superjson#using-with-nextjs).

**Comment by chernodub** [comment]
Faced the same issue. I believe it would be a good idea to abstract such stuff from the users' code by swapping the default serializer. I proposed the idea before https://github.com/vercel/next.js/discussions/45702

**Comment by joulev** [comment]
As a follow up, I just had a mind-blowing discovery that the following works in prod mode

```tsx
export default function Home(props) {
  console.log(props);
  return <div>Hello world</div>;
}

export function getServerSideProps() { // or getStaticProps
  return {
    props: { date: new Date() },
  };
}
```

`props` will be `{ date: string }` as expected, without any errors with `next build && next start`. On `next dev` it is checked however.
